### PR TITLE
Eliminate boxing when entering critical sections

### LINKED ISSUE #21229: Remove allocation from ConcurrencyDetector critical section management
For every entity we read [in QueryingEnumerable](https://github.com/dotnet/efcore/blob/6242987ed5c66e25dcac78281059b54dae3e33f1/src/EFCore.Relational/Query/Internal/QueryingEnumerable.cs#L136), we incur a heap allocation because EnterCriticalSection returns an IDisposable. Returning the struct directly would avoid that.

This would probably mean making ConcurrencyDetector an internal implementation detail of QueryingEnumerable, and not a replaceable service.

### PR DESCRIPTION
**Pull request by roji**
Closes #21229

### CODE REVIEW

File: src/EFCore/Infrastructure/IConcurrencyDetector.cs [hunk 1]
```diff
@@ -21,9 +21,14 @@ namespace Microsoft.EntityFrameworkCore.Infrastructure
     public interface IConcurrencyDetector
     {
         /// <summary>
-        ///     Call to enter the critical section.
+        ///     Enter a critical section.
```
Comment (AndriySvyryd) [inline src/EFCore/Infrastructure/IConcurrencyDetector.cs:24]:
```suggestion
        ///     Enters a critical section.
```

File: src/EFCore/Infrastructure/ConcurrencyDetectorCriticalSectionDisposer.cs [hunk 2]
```diff
@@ -0,0 +1,28 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+
+namespace Microsoft.EntityFrameworkCore.Infrastructure
+{
+    /// <summary>
+    /// A <see cref="IDisposable" /> returned by an <see cref="IConcurrencyDetector" />, which will exit the ongoing
+    /// critical section when disposed.
```
Comment (smitpatel) [inline src/EFCore/Infrastructure/ConcurrencyDetectorCriticalSectionDisposer.cs:10]:
```suggestion
    ///     A <see cref="IDisposable" /> returned by an <see cref="IConcurrencyDetector" />, which will exit the ongoing
    ///     critical section when disposed.
```

**Comment by smitpatel** [comment]
no tests? 😢

**Comment by roji** [comment]
@smitpatel I think all the existing ones in ConcurrencyDetectorTestBase are valid, no? Or you mean you want explicit testing for ExitCriticalSection?

File: src/EFCore/Infrastructure/ConcurrencyDetectorCriticalSectionDisposer.cs [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+
+using System;
+
+namespace Microsoft.EntityFrameworkCore.Infrastructure
+{
+    /// <summary>
+    ///     A <see cref="IDisposable" /> returned by an <see cref="IConcurrencyDetector" />, which will exit the ongoing
+    ///     critical section when disposed.
+    /// </summary>
+    public class ConcurrencyDetectorCriticalSectionDisposer : IDisposable
```
Comment (AndriySvyryd) [inline src/EFCore/Infrastructure/ConcurrencyDetectorCriticalSectionDisposer.cs:12]: This isn't a struct, so we will still be allocating. What's the purpose of this change?

Comment (roji) [reply src/EFCore/Infrastructure/ConcurrencyDetectorCriticalSectionDisposer.cs:12]: Sorry, that was a typo (it's 3AM here). Changing to readonly struct.

Comment (roji) [reply src/EFCore/Infrastructure/ConcurrencyDetectorCriticalSectionDisposer.cs]: (maybe concurrency-related programming at 3AM isn't such a good idea?)

Comment (AndriySvyryd) [reply src/EFCore/Infrastructure/ConcurrencyDetectorCriticalSectionDisposer.cs]: I though it was an intentional bug like what Arthur suggested to include in every PR

File: src/EFCore/Internal/ConcurrencyDetector.cs [hunk 4]
```diff
@@ -60,10 +51,16 @@ public virtual IDisposable EnterCriticalSection()
             }
 
             _refCount++;
-            return _disposer;
+            return new ConcurrencyDetectorCriticalSectionDisposer(this);
```
Comment (smitpatel) [inline src/EFCore/Internal/ConcurrencyDetector.cs:54]: Does this not create a new struct every time? (rather than reusing like how previous did)

Comment (AndriySvyryd) [reply src/EFCore/Internal/ConcurrencyDetector.cs:54]: struct construction is cheap

Comment (roji) [reply src/EFCore/Internal/ConcurrencyDetector.cs:54]: It does, but creating a struct (on the stack) is free - especially as here it's just wrapping a pointer. So it's the same as allocating a long, or a reference on the stack.

**Comment by roji** [comment]
Just for fun, here's the benchmark to show this:

``` ini

BenchmarkDotNet=v0.12.0, OS=ubuntu 20.04
Intel Xeon W-2133 CPU 3.60GHz, 1 CPU, 12 logical and 6 physical cores
.NET Core SDK=5.0.100-preview.6.20266.3
  [Host]     : .NET Core 3.1.1 (CoreCLR 4.700.19.60701, CoreFX 4.700.19.60801), X64 RyuJIT
  DefaultJob : .NET Core 3.1.1 (CoreCLR 4.700.19.60701, CoreFX 4.700.19.60801), X64 RyuJIT


```
|       Method |      Mean |     Error |    StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|------------- |----------:|----------:|----------:|-------:|------:|------:|----------:|
| MyDisposable | 0.0226 ns | 0.0030 ns | 0.0027 ns |      - |     - |     - |         - |
|   Disposable | 6.4407 ns | 0.2355 ns | 0.2202 ns | 0.0056 |     - |     - |      24 B |

<details>
<summary>Benchmark code</summary>

```c#
[MemoryDiagnoser]
public class Program
{
    [Benchmark]
    public void MyDisposable()
    {
        using var _ = GetMyDisposable();
    }

    [Benchmark]
    public void Disposable()
    {
        using var _ = GetDisposable();
    }

    static IDisposable GetDisposable() => new MyDisposable();

    static MyDisposable GetMyDisposable() => new MyDisposable();

    static void Main(string[] args)
        => BenchmarkRunner.Run<Program>();
}

readonly struct MyDisposable : IDisposable
{
    public void Dispose() {}
}
```
</details>

Now am off to fix the test failures...

File: src/EFCore/Infrastructure/ConcurrencyDetectorCriticalSectionDisposer.cs (hunk 3 above)
Comment (roji) [reply src/EFCore/Infrastructure/ConcurrencyDetectorCriticalSectionDisposer.cs]: Ah of course, that's what it was. Congrats for winning the prize @AndriySvyryd!
